Creating newspaper issues is slow and fails with OutOfMemoryError

Creating newspaper issues takes 10 seconds per process on an idle Linux (JVM has 4 GB RAM) and failed after hours with `OutOfMemoryError` because the module loader couldn’t unzip a module in memory to look if it is the Command Module. There is something wrong with the caching strategy of modules. The module was used 7 hours so there shouldn’t be any look-up at all because the module has already been loaded.

```
[ERROR] 2021-02-24 18:10:26.194 [Erzeuge Zeitungsvorgänge: Tydzno_017986109] EmptyTask - java.io.IOException: java.lang.reflect.InvocationTargetException
java.lang.IllegalStateException: java.io.IOException: java.lang.reflect.InvocationTargetException
        at org.apache.catalina.webresources.AbstractSingleArchiveResourceSet.getArchiveEntry(AbstractSingleArchiveResourceSet.java:101) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.AbstractArchiveResourceSet.getResource(AbstractArchiveResourceSet.java:251) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.StandardRoot.getResourcesInternal(StandardRoot.java:328) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.CachedResource.validateResources(CachedResource.java:158) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.Cache.getResources(Cache.java:147) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.StandardRoot.getResources(StandardRoot.java:316) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.StandardRoot.getClassLoaderResources(StandardRoot.java:232) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.loader.WebappClassLoaderBase.findResources(WebappClassLoaderBase.java:991) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.loader.WebappClassLoaderBase.getResources(WebappClassLoaderBase.java:1086) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1196) ~[?:?]
        at java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1221) ~[?:?]
        at java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1265) ~[?:?]
        at java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1300) ~[?:?]
        at java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1385) ~[?:?]
        at org.kitodo.serviceloader.KitodoServiceLoader.loadModule(KitodoServiceLoader.java:100) ~[kitodo-api-3.2.1-SNAPSHOT.jar:?]
        at org.kitodo.production.services.command.CommandService.runCommand(CommandService.java:50) ~[classes/:3.2.1-SNAPSHOT]
        at org.kitodo.production.services.command.CommandService.runCommand(CommandService.java:77) ~[classes/:3.2.1-SNAPSHOT]
        at org.kitodo.production.services.file.FileService.createMetaDirectory(FileService.java:122) ~[classes/:3.2.1-SNAPSHOT]
        at org.kitodo.production.services.file.FileService.createProcessFolders(FileService.java:234) ~[classes/:3.2.1-SNAPSHOT]
        at org.kitodo.production.services.file.FileService.createProcessLocation(FileService.java:214) ~[classes/:3.2.1-SNAPSHOT]
        at org.kitodo.production.process.NewspaperProcessesGenerator.createMetadataFileForProcess(NewspaperProcessesGenerator.java:582) ~[classes/:3.2.1-SNAPSHOT]
        at org.kitodo.production.process.NewspaperProcessesGenerator.createProcess(NewspaperProcessesGenerator.java:522) ~[classes/:3.2.1-SNAPSHOT]
        at org.kitodo.production.process.NewspaperProcessesGenerator.nextStep(NewspaperProcessesGenerator.java:297) ~[classes/:3.2.1-SNAPSHOT]
        at org.kitodo.production.helper.tasks.GeneratesNewspaperProcessesThread.run(GeneratesNewspaperProcessesThread.java:80) ~[classes/:3.2.1-SNAPSHOT]
Caused by: java.io.IOException: java.lang.reflect.InvocationTargetException
        at org.apache.tomcat.util.compat.Jre9Compat.jarFileNewInstance(Jre9Compat.java:236) ~[tomcat9-util-9.0.31.jar:9.0.31]
        at org.apache.tomcat.util.compat.JreCompat.jarFileNewInstance(JreCompat.java:162) ~[tomcat9-util-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.AbstractArchiveResourceSet.openJarFile(AbstractArchiveResourceSet.java:307) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.AbstractSingleArchiveResourceSet.getArchiveEntry(AbstractSingleArchiveResourceSet.java:97) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        ... 23 more
Caused by: java.lang.reflect.InvocationTargetException
        at jdk.internal.reflect.GeneratedConstructorAccessor7.newInstance(Unknown Source) ~[?:?]
        at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:490) ~[?:?]
        at org.apache.tomcat.util.compat.Jre9Compat.jarFileNewInstance(Jre9Compat.java:233) ~[tomcat9-util-9.0.31.jar:9.0.31]
        at org.apache.tomcat.util.compat.JreCompat.jarFileNewInstance(JreCompat.java:162) ~[tomcat9-util-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.AbstractArchiveResourceSet.openJarFile(AbstractArchiveResourceSet.java:307) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.AbstractSingleArchiveResourceSet.getArchiveEntry(AbstractSingleArchiveResourceSet.java:97) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        ... 23 more
Caused by: java.lang.OutOfMemoryError: Java heap space
        at java.util.zip.ZipFile$Source.initCEN(ZipFile.java:1510) ~[?:?]
        at java.util.zip.ZipFile$Source.<init>(ZipFile.java:1288) ~[?:?]
        at java.util.zip.ZipFile$Source.get(ZipFile.java:1251) ~[?:?]
        at java.util.zip.ZipFile$CleanableResource.<init>(ZipFile.java:732) ~[?:?]
        at java.util.zip.ZipFile$CleanableResource.get(ZipFile.java:849) ~[?:?]
        at java.util.zip.ZipFile.<init>(ZipFile.java:247) ~[?:?]
        at java.util.zip.ZipFile.<init>(ZipFile.java:177) ~[?:?]
        at java.util.jar.JarFile.<init>(JarFile.java:348) ~[?:?]
        at jdk.internal.reflect.GeneratedConstructorAccessor7.newInstance(Unknown Source) ~[?:?]
        at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:490) ~[?:?]
        at org.apache.tomcat.util.compat.Jre9Compat.jarFileNewInstance(Jre9Compat.java:233) ~[tomcat9-util-9.0.31.jar:9.0.31]
        at org.apache.tomcat.util.compat.JreCompat.jarFileNewInstance(JreCompat.java:162) ~[tomcat9-util-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.AbstractArchiveResourceSet.openJarFile(AbstractArchiveResourceSet.java:307) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.AbstractSingleArchiveResourceSet.getArchiveEntry(AbstractSingleArchiveResourceSet.java:97) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.AbstractArchiveResourceSet.getResource(AbstractArchiveResourceSet.java:251) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.StandardRoot.getResourcesInternal(StandardRoot.java:328) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.CachedResource.validateResources(CachedResource.java:158) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.Cache.getResources(Cache.java:147) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.StandardRoot.getResources(StandardRoot.java:316) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.webresources.StandardRoot.getClassLoaderResources(StandardRoot.java:232) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.loader.WebappClassLoaderBase.findResources(WebappClassLoaderBase.java:991) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at org.apache.catalina.loader.WebappClassLoaderBase.getResources(WebappClassLoaderBase.java:1086) ~[tomcat9-catalina-9.0.31.jar:9.0.31]
        at java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1196) ~[?:?]
        at java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1221) ~[?:?]
        at java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1265) ~[?:?]
        at java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1300) ~[?:?]
        at java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1385) ~[?:?]
        at org.kitodo.serviceloader.KitodoServiceLoader.loadModule(KitodoServiceLoader.java:100) ~[kitodo-api-3.2.1-SNAPSHOT.jar:?]
        at org.kitodo.production.services.command.CommandService.runCommand(CommandService.java:50) ~[classes/:3.2.1-SNAPSHOT]
        at org.kitodo.production.services.command.CommandService.runCommand(CommandService.java:77) ~[classes/:3.2.1-SNAPSHOT]
        at org.kitodo.production.services.file.FileService.createMetaDirectory(FileService.java:122) ~[classes/:3.2.1-SNAPSHOT]
```

--- a/Kitodo/src/main/java/org/kitodo/production/services/command/CommandService.java
+++ b/Kitodo/src/main/java/org/kitodo/production/services/command/CommandService.java
@@ -27,9 +27,16 @@ import org.kitodo.production.services.ServiceManager;
 import org.kitodo.serviceloader.KitodoServiceLoader;
 
 public class CommandService {
-    private static final CommandService commandService = ServiceManager.getCommandService();
+    private final CommandInterface commandModule;
     private final ArrayList<CommandResult> finishedCommandResults = new ArrayList<>();
     private final Random random = new Random(1000000);
+
+    /**
+     * Initialize Command Service.
+     */
+    public CommandService() {
+        commandModule = new KitodoServiceLoader<CommandInterface>(CommandInterface.class).loadModule();
+    }
 
     /**
      * Method executes a script string.
@@ -47,7 +54,7 @@ public class CommandService {
         if (Objects.isNull(script)) {
             return null;
         }
-        CommandResult commandResult = commandService.runCommand(script);
+        CommandResult commandResult = commandModule.runCommand(random.nextInt(), script);
         List<String> commandResultMessages = commandResult.getMessages();
         if (!commandResultMessages.isEmpty() && commandResultMessages.get(0).contains("IOException")) {
             throw new IOException(commandResultMessages.get(1));

--- a/Kitodo/src/main/java/org/kitodo/production/services/command/CommandService.java
+++ b/Kitodo/src/main/java/org/kitodo/production/services/command/CommandService.java
@@ -30,6 +30,7 @@ public class CommandService {
     private static final CommandService commandService = ServiceManager.getCommandService();
     private final ArrayList<CommandResult> finishedCommandResults = new ArrayList<>();
     private final Random random = new Random(1000000);
+
     /**
      * Method executes a script string.
      *

--- a/Kitodo/src/main/java/org/kitodo/production/services/command/CommandService.java
+++ b/Kitodo/src/main/java/org/kitodo/production/services/command/CommandService.java
@@ -23,13 +23,13 @@ import java.util.Random;
 
 import org.kitodo.api.command.CommandInterface;
 import org.kitodo.api.command.CommandResult;
+import org.kitodo.production.services.ServiceManager;
 import org.kitodo.serviceloader.KitodoServiceLoader;
 
 public class CommandService {
-
-    private final ArrayList<CommandResult> finishedCommandResults = new ArrayList<>();
+	private static final CommandService commandService = ServiceManager.getCommandService();
+	private final ArrayList<CommandResult> finishedCommandResults = new ArrayList<>();
     private final Random random = new Random(1000000);
-
     /**
      * Method executes a script string.
      *
@@ -46,10 +46,7 @@ public class CommandService {
         if (Objects.isNull(script)) {
             return null;
         }
-        KitodoServiceLoader<CommandInterface> serviceLoader = new KitodoServiceLoader<>(CommandInterface.class);
-        CommandInterface command = serviceLoader.loadModule();
-
-        CommandResult commandResult = command.runCommand(random.nextInt(), script);
+        CommandResult commandResult = commandService.runCommand(script);
         List<String> commandResultMessages = commandResult.getMessages();
         if (!commandResultMessages.isEmpty() && commandResultMessages.get(0).contains("IOException")) {
             throw new IOException(commandResultMessages.get(1));

--- a/Kitodo/src/main/java/org/kitodo/production/services/command/CommandService.java
+++ b/Kitodo/src/main/java/org/kitodo/production/services/command/CommandService.java
@@ -27,8 +27,8 @@ import org.kitodo.production.services.ServiceManager;
 import org.kitodo.serviceloader.KitodoServiceLoader;
 
 public class CommandService {
-	private static final CommandService commandService = ServiceManager.getCommandService();
-	private final ArrayList<CommandResult> finishedCommandResults = new ArrayList<>();
+    private static final CommandService commandService = ServiceManager.getCommandService();
+    private final ArrayList<CommandResult> finishedCommandResults = new ArrayList<>();
     private final Random random = new Random(1000000);
     /**
      * Method executes a script string.

